core/src/main/java/org/elasticsearch/script/TemplateService.java

```diff
@@ -0,0 +1,239 @@
+    private final Backend backend;
+    private final ScriptPermits scriptPermits;
+    private final CachingCompiler<CacheKey> compiler;
+
+    public TemplateService(Settings settings, Environment env,
+            ResourceWatcherService resourceWatcherService, Backend backend,
+            ScriptContextRegistry scriptContextRegistry, ScriptSettings scriptSettings,
+            ScriptMetrics scriptMetrics) throws IOException {
+        Objects.requireNonNull(scriptContextRegistry);
+
+        this.backend = backend;
+        this.scriptPermits = new ScriptPermits(settings, scriptSettings, scriptContextRegistry);
+        this.compiler = new CachingCompiler<CacheKey>(settings, env, resourceWatcherService,
+                scriptMetrics, "template") {
+            @Override
+            protected CacheKey cacheKeyForFile(String baseName, String extension) {
+                if (false == backend.getType().equals(extension)) {
+                    /* For backwards compatibility templates are in the scripts directory and we
+                     * must ignore everything but templates. */
+                    return null;
+                }
+                return new CacheKey(baseName, DEFAULT_CONTENT_TYPE);
+            }
+
+            @Override
+            protected CacheKey cacheKeyFromClusterState(StoredScriptSource scriptMetadata) {
+                String contentType = DEFAULT_CONTENT_TYPE;
+                if (scriptMetadata.getOptions() != null) {
+                    contentType = scriptMetadata.getOptions().get(Script.CONTENT_TYPE_OPTION);
+                    if (contentType == null) {
+                        contentType = DEFAULT_CONTENT_TYPE;
+                    }
+                }
+                return new CacheKey(scriptMetadata.getCode(), contentType);
+            }
+
+            @Override
+            protected StoredScriptSource lookupStoredScript(ClusterState clusterState,
+                    CacheKey cacheKey) {
+                if (clusterState == null) {
+                    return null;
+                }
+                MetaData metaData = clusterState.metaData();
+                if (metaData == null) {
+                    return null;
+                }
+                ScriptMetaData scriptMetaData = clusterState.metaData().custom(ScriptMetaData.TYPE);
+                if (scriptMetaData == null) {
+                    return null;
+                }
+
+                /* This process throws away which is fine because you aren't allowed to specify it
+                 * when using a stored template anyway. */
+                String id = cacheKey.idOrCode;
+                // search template requests can possibly pass in the entire path instead
+                // of just an id for looking up a stored script, so we parse the path and
+                // check for appropriate errors
+                String[] path = id.split("/");
+
+                if (path.length == 3) {
```

Comment:
Hmmm... I 'm fairly sure this part isn't actually need in `ScriptService`. Some of it is. I can pull that into a method on the parent.